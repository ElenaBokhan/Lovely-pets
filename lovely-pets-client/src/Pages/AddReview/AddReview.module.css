.addReviewPage {
    padding: 0 120px;
}

.reviewTextarea {
    height: 88px;
    padding: 14px 16px;
    border-radius: 12px;
    border: 1px solid #CFD8DC;
}

.field {
    display: flex;
    gap: 40px;
    margin: 40px 0;
}

.field span {
    width: 30%;
}

.field input {
    width: 70%;
}

@media (width <= 576px) {
    .addReviewPage {
        padding: 0;
    }

    .form {
        padding: 0 20px;
        text-align: center;
        display: flex;
        flex-direction: column;
    }

    .field{
        flex-direction: column;
    }

    .field span {
        width: 100%
    }

    .field input {
        width: 100%;
    }

    .stars {
        display: flex;
        gap: 20px;
        justify-content: center;
    }
}