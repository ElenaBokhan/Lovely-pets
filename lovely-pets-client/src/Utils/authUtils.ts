import {ILoginResponse} from 'Api/authClient';

export const setLoginToStorage = (loginData: ILoginResponse) => {
    const {access_token, refresh_token, username, id} = loginData;

    localStorage.setItem('access_token', access_token);
    localStorage.setItem('refresh_token', refresh_token);
    localStorage.setItem('userName', username);
    localStorage.setItem('userId', id);
};

export const removeLoginToStorage = () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('userName');
    localStorage.removeItem('userId');
};
