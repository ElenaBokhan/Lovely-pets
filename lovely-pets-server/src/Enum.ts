export enum ESortFilter {
    POPULAR = 'POPULAR',
    NEW = 'NEW',
    CHEAPER = 'CHEAPER',
    EXPENSIVE = 'EXPENSIVE',
    RATING = 'RATING',
    DISCOUNT = 'DISCOUNT',
}
