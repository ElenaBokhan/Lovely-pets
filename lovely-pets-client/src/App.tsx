import './index.css';
import {Layout} from 'Components/Layout/Layout';
import {Catalog} from 'Pages/Catalog/Catalog';
import {Favourites} from 'Pages/Favourites/Favourites';
import {PageNotFound} from 'Pages/PageNotFound/PageNotFound';
import {PetCard} from 'Pages/PetCard/PetCard';
import {Profile} from 'Pages/Profile/Profile';
import {RouterProvider, createBrowserRouter, RouteObject} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import {SignUp} from 'Pages/SignUp/SignUp';
import {SignIn} from 'Pages/SignIn/SignIn';
import {Cart} from 'Pages/Cart/Cart';
import {AddReview} from 'Pages/AddReview/AddReview';
import {AddProduct} from 'Pages/AddProduct/AddProduct';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {withTranslation} from 'react-i18next';
import {useEffect} from 'react';
import {setCartFromStorage, setCartToStorage} from 'Store/cartStore';
import 'Style/Themes.module.css';
import Theme from 'Components/Theme/Theme';

export const router: RouteObject[] = [
    {
        path: '/',
        element: <Layout />,
        errorElement: <div />, // TODO: добавить errorBoundary, пока заглушка
        children: [
            {
                path: '/',
                element: <Catalog />,
            },
            {
                path: '/favourites',
                element: <Favourites />,
            },
            {
                path: '/pets/:petId',
                element: <PetCard />,
            },
            {
                path: '/profile',
                element: <Profile />,
            },
            {
                path: '/addReview',
                element: <AddReview />,
            },
            {
                path: '/addPet',
                element: <AddProduct />,
            },
            {
                path: '/signup',
                element: <SignUp />,
            },
            {
                path: '/signin',
                element: <SignIn />,
            },
            {
                path: '/cart',
                element: <Cart />,
            },
            {
                path: '*',
                element: <PageNotFound />,
            },
        ],
    },
];

export const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: false,
        },
    },
});

const App = () => {
    useEffect(() => {
        window.addEventListener('load', setCartFromStorage);
        window.addEventListener('beforeunload', setCartToStorage);
        return () => {
            window.removeEventListener('load', setCartFromStorage);
            window.removeEventListener('beforeunload', setCartToStorage);
        };
    }, []);

    return (
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={createBrowserRouter(router)} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );
};

export default withTranslation()(App);
