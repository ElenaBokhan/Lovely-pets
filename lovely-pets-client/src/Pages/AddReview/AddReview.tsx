import {Button} from 'Components/Common/Button/Button';
import {Text} from 'Components/Common/Text/Text';
import {TitlePage} from 'Components/Common/TitlePage/TitlePage';
import styles from 'Pages/AddReview/AddReview.module.css';
import {useLocation, useNavigate} from 'react-router-dom';
import starFillIcon from 'assets/ic-star-fill.svg';
import starIcon from 'assets/ic-star.svg';
import {useState} from 'react';
import {ETestId} from 'Enum';
import {IconButton} from 'Components/Common/IconButton/IconButton';
import {Gap} from 'Components/Common/Gap/Gap';
import {useAddReviewMutation} from 'Api/queryHooks';
import {useTranslation} from 'react-i18next';

export const AddReview = () => {
    const {state} = useLocation();
    const userId = localStorage.getItem('userId');
    const userName = localStorage.getItem('userName');
    const {t} = useTranslation();

    const {mutateAsync: addReview} = useAddReviewMutation();
    const navigate = useNavigate();

    const [reviewText, setReviewText] = useState<string>('');
    const [currentRating, setCurrentRating] = useState<number>(null);

    const handleReviewChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
        const reviewText = event.currentTarget.value;
        setReviewText(reviewText);
    };

    const handleSubmit = async (event: React.SyntheticEvent<HTMLFormElement>) => {
        event.preventDefault();

        if (currentRating || reviewText) {
            const updatedPet: IPet = await addReview({
                petId: state?._id,
                rating: currentRating,
                text: reviewText,
                userId,
                userName,
            });
            navigate(`/pets/${updatedPet._id}`);
        }
    };

    const handleSetRating = (rate: number) => {
        setCurrentRating(rate);
    };

    const renderStars = () => {
        return (
            <div className={styles.stars}>
                {[...Array(5)].map((_, index) => (
                    <IconButton
                        onClick={() => handleSetRating(index + 1)}
                        key={index}
                        icon={index < currentRating ? starFillIcon : starIcon}
                        alt={'ratingProduct'}
                    />
                ))}
            </div>
        );
    };

    const renderAddReviewForm = () => {
        return (
            <form className={styles.form} onSubmit={handleSubmit}>
                <div className={styles.field}>
                    <Text value={t('pages.review.rating')} />
                    {renderStars()}
                </div>
                <div className={styles.field}>
                    <Text value={t('pages.review.comment.label')} />
                    <input
                        className={styles.reviewTextarea}
                        type="textarea"
                        data-test-id={ETestId.ADD_REVIEW_TEXT_INPUT}
                        onChange={handleReviewChange}
                        placeholder={t('pages.review.comment.placeholder')}
                        value={reviewText}
                    />
                </div>
                <Button
                    testId={ETestId.ADD_REVIEW_SUBMIT_BUTTON}
                    type="submit"
                    label={t('pages.review.submit_button')}
                />
            </form>
        );
    };

    return (
        <div className={styles.addReviewPage}>
            <TitlePage label={t('pages.review.title', {name: state?.name})} />
            <Gap size={40} />
            <hr />
            {renderAddReviewForm()}
        </div>
    );
};
