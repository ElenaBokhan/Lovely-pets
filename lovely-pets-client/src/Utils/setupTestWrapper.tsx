import {queryClient, router} from 'App';
import {render} from '@testing-library/react';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import {QueryClientProvider} from '@tanstack/react-query';
import Theme from 'Components/Theme/Theme';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={memoryRouter} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );

    return memoryRouter;
};
