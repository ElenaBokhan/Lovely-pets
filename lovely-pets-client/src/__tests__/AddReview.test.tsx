import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {render, screen} from '@testing-library/react';
import {AddReview} from 'Pages/AddReview/AddReview';
import {BrowserRouter, createMemoryRouter} from 'react-router-dom';
import * as ReactRouter from 'react-router';
import {ETestId} from 'Enum';
import {userEvent} from '@testing-library/user-event';
import {queryClient} from 'App';
import {QueryClientProvider} from '@tanstack/react-query';

describe('Тесты для компонента AddReview', () => {
    let router: ReturnType<typeof createMemoryRouter> | undefined;
    let submitBtn: HTMLButtonElement | undefined;
    let textInput: HTMLInputElement | undefined;
    const name = 'productName';
    const productId = '123';

    beforeEach(() => {
        jest.spyOn(ReactRouter, 'useLocation').mockImplementation(() => ({
            state: {name, _id: productId},
            pathname: '',
            hash: '',
            key: '',
            search: '',
        }));
        router = setupTestsWrapper({initialRouterEntries: ['/addReview']});
        submitBtn = screen.getByTestId(ETestId.ADD_REVIEW_SUBMIT_BUTTON);
        textInput = screen.getByTestId(ETestId.ADD_REVIEW_TEXT_INPUT);
    });

    describe('Проверяем элементы компонента AddReview', () => {
        test('Проверяем что поле текстового сообщения и кнопка добавления отзыва доступны', async () => {
            expect(submitBtn).toBeEnabled();
            expect(textInput).toBeEnabled();
        });
        test('Проверяем что запрос не выполняется при незаполненных поле текстового сообщения и редиректа не происходит', async () => {
            expect(textInput).toHaveDisplayValue('');

            await userEvent.click(submitBtn);
            expect(router?.state.location.pathname).toEqual('/addReview');
        });
    });

    describe('Снапшот компонента AddReview', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <QueryClientProvider client={queryClient}>
                    <BrowserRouter>
                        <AddReview />
                    </BrowserRouter>
                </QueryClientProvider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});
