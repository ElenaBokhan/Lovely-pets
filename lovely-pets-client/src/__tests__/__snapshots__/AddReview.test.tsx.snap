// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента AddReview Снапшот компонента AddReview Snapshot 1`] = `
<DocumentFragment>
  <div
    class="addReviewPage"
  >
    <div
      data-testid="titlePage"
    >
      <a
        class="breadcrumbs"
        data-testid="breadcrumbsButton"
        href="/"
      >
        <img
          alt="arrow"
          src="SvgrURL"
        />
        <span
          class="grey undefined subscribe2"
        >
          Назад
        </span>
      </a>
      <span
        class="undefined undefined head1"
      >
        Отзыв о товаре productName
      </span>
    </div>
    <div
      style="height: 40px;"
    />
    <hr />
    <form>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Общая оценка
        </span>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="SvgrURL"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="SvgrURL"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="SvgrURL"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="SvgrURL"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="SvgrURL"
          />
        </button>
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Комментарий
        </span>
        <input
          class="reviewTextarea"
          data-test-id="addReviewTextInput"
          placeholder="Поделитесь впечатлениями о товаре"
          type="textarea"
          value=""
        />
      </div>
      <button
        class="button standard"
        data-test-id="addReviewSubmitButton"
        data-value="Отправить отзыв"
        type="submit"
      >
        Отправить отзыв
      </button>
    </form>
  </div>
</DocumentFragment>
`;
